README.md
Object Persistence In C (Pre-Alpha)
===================================

OPIC is a new approach to serialize general data structures, object types, and
primitive types. It's a new data stack from the ground up -- from object
oriented model, generics, to memory management. With everything redesigned for
scalability, we visioned a new distributed computing ecosystem for the 21st
century.

Our key to success is to extend the memory oriented programming model to other
tier of storages: disk, SSD, network, or even tape. Accessing data should not be
limited by memory address on single machine; instead, all data types should be
serializable so that it can be shared across network or save for future use.
Having serialzability as the core design brings many benefits for free:

* Scalability. When all the data types are serializable, the application can
easily scale up with multiple strategies, including sharing data between nodes,
off load certain data for later use, or build caches from the same serialization
abstraction.

* Data persistence. Think of databases' durability. Since all the data types are
serializable, you can save the snapshot of the entire program state whenever you
want. This can make database applications way easier to build, and also make
debugging easier.

* Generality for data processing applications. All the data processing systems,
including RDBMS, map reduce, search engines, even RPC involves are some sort of
special case of serialization. Why duplicates the effort for each of those
applications instead of having one powerful and optimized framework?

Most of the functionalities will be implemented in C. Our end goal is to embed
this framework into other higher level languages like python, java, nodejs for
wider adoption.

To begin with, please check out our [documentation page][doc].

[doc]: http://dryman.github.com/opic/

LICENSE
-------

Copyright (c) 2016 Felix Chern

    OPIC is free software: you can redistribute it and/or modify it under the
    terms of the GNU Lesser General Public License as published by the Free
    Software Foundation, either version 3 of the License, or (at your option)
    any later version.

    OPIC is distributed in the hope that it will be useful, but WITHOUT ANY
    WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
    FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for
    more details.

    You should have received a copy of the GNU Lesser General Public License
    along with OPIC.  If not, see <http://www.gnu.org/licenses/>.

In future we may also form a commercial license and support.

CODE STYLE FOR CONTRIBUTORS
---------------------------

Please follow the [GNU Coding Standards][gnuc]. On the moment of this
writing we still have some portion of the code hasn't yet convert to GNU
formatting styles, but this should be fixed in near future.

[gnuc]: https://www.gnu.org/prep/standards/standards.html



p99/ChangeLog
commit d83e6ac40e442ba427bb1273e4c2d9e88428f742
Author: Jens Gustedt <Jens.Gustedt@inria.fr>
Date:   Sun Jan 31 11:31:04 2016 +0100

    implement a condition-mutex

p99/p99_cm.h

commit 6af85ab4fbf33169a0cccc0d899491570d8fa50a
Author: Jens Gustedt <Jens.Gustedt@inria.fr>
Date:   Sun Jan 31 11:30:01 2016 +0100

    minor documentation change

p99/p99_count.h

commit 4536e935303a0ef8346cbb23d226cb146990c2fd
Author: Jens Gustedt <Jens.Gustedt@inria.fr>
Date:   Tue Sep 22 14:59:37 2015 +0000

    avoid another usesless case of testing for qualifiers

p99/p99_generic.h

commit db0f0628f63faee96bc10c54dd65084c125a9253
Author: Jens Gustedt <Jens.Gustedt@inria.fr>
Date:   Tue Sep 8 06:10:09 2015 +0000

    avoid some combinatorial explosion caused by qualifiers in _Generic

p99/p99_generic.h
p99/p99_int.h

commit ed45a527b92c2037793bc97df37f6ce14bf311c2
Author: (no author) <(no author)@no.where>
Date:   Tue Jun 9 09:05:55 2015 +0200

    whitespace and copyright update

p99/p99_atomic.h
p99/p99_block.h
p99/p99_c99.h
p99/p99_callback.h
p99/p99_clib.h
p99/p99_compiler.h
p99/p99_constraint.h
p99/p99_for.h
p99/p99_futex_c11.h
p99/p99_futex_linux.h
p99/p99_generated.h
p99/p99_generic.h
p99/p99_getopt.h
p99/p99_libc.h
p99/p99_map.h
p99/p99_notifier.h
p99/p99_rand.h
p99/p99_rwl.h
p99/p99_threads.h
p99/p99_threads_posix.h
p99/p99_tp.h
p99/p99_try.h
p99/p99_typenames.h

commit dd41b304ccfb4bfd1ce4bf1b8efc392c4054fe6f
Author: Jens Gustedt <Jens.Gustedt@inria.fr>
Date:   Tue Jun 9 08:53:09 2015 +0200

    add DESCENDING
    
    also upgrade the findSvnAuthors script to changes in astyle

p99/Makefile
p99/p99_generated.h
scripts/findSvnAuthors
scripts/p99_generate.pl

commit 2c4b05c0f0c517b3ff87c6d8984d8dd0c7966067
Author: Jens Gustedt <Jens.Gustedt@inria.fr>
Date:   Mon Feb 16 22:09:57 2015 +0100

    OS X uses a noreturn identifier where they shouldn't
    
    Reported and tested by: Francisco G. <frangio.1@gmail.com>

p99/p99_compiler.h

commit ac46c9927ae45363181c21c63e4a0838d60d9c50
Author: Jens Gustedt <Jens.Gustedt@inria.fr>
Date:   Sun Dec 14 18:12:08 2014 +0100

    remove unused local variables

p99/p99_atomic.h

commit 83cb84c05701a6c203006fb3752324a11b05e27b
Author: Jens Gustedt <Jens.Gustedt@inria.fr>
Date:   Sun Dec 14 18:11:34 2014 +0100

    add some information about glibc C11 support

p99/p99_libc.h

commit ae565ed901474a99156c019be32b7181011dbdcb
Author: Jens Gustedt <Jens.Gustedt@inria.fr>
Date:   Fri Oct 3 14:46:02 2014 +0200

    remove support for ssize_t, this is POSIX' business

p99/p99_typenames.h

commit 106b1ec19c8cca5ef4382bf0b12d92f52bb2f3d1
Author: Jens Gustedt <Jens.Gustedt@inria.fr>
Date:   Fri Oct 3 14:06:19 2014 +0200

    p99_try.h needs math.h for HUGE_VALF

p99/p99_try.h
tests/test-p99-error.c

commit f637b688010fd3956f19786b90d81146c929c7a3
Author: Jens Gustedt <Jens.Gustedt@inria.fr>
Date:   Fri Oct 3 13:44:25 2014 +0200

    some missing bits that are necessary after the rearangement of C11
    library additions

p99/p99_rand.h
tests/Makefile
tests/test-p99-c11.c

commit 0ad97a9a3532d4915a3352d656f970c5f8797b9b
Author: Jens Gustedt <Jens.Gustedt@inria.fr>
Date:   Fri Aug 15 09:15:30 2014 +0200

    improve the thread test to mix cnd_signal and cnd_broadcast

tests/test-p99-thread.c

commit 4c26b2d5ce3b612ebaa093c9b9df281f9c0eb1c5
Author: Jens Gustedt <Jens.Gustedt@inria.fr>
Date:   Fri Aug 15 09:14:51 2014 +0200

    add a newline in the error report

p99/p99_threads.h

commit 9a73aa2220e4cd2b5d98d59131a9985bbf051c98
Author: Jens Gustedt <Jens.Gustedt@inria.fr>
Date:   Fri Aug 15 09:14:26 2014 +0200

    add prototype for syscall

p99/p99_futex_linux.h

commit 2342c336aa71fac8e96e9605bed1a2e056e3c3cd
Author: Jens Gustedt <Jens.Gustedt@inria.fr>
Date:   Thu Jul 31 16:28:33 2014 +0200

    refine the p99_rwl implementation
    
    this new version will only call wakeup if there is a waiter
    
    so a typically fast path for a non-congested lock will never do a futex
    call, but do all with atomics

p99/p99_rwl.h
tests/libp99-weak.c

commit ea6675d144dad18cb58fc5a0589b630fba38d67b
Author: Jens Gustedt <Jens.Gustedt@inria.fr>
Date:   Thu Jul 31 16:00:35 2014 +0200

    add a read-write lock structure based on futex

p99/p99_rwl.h

commit 60f5ab44fe1a71189f43e6bafe92358cb522771f
Author: Jens Gustedt <Jens.Gustedt@inria.fr>
Date:   Thu Jul 31 11:27:13 2014 +0200

    separate the C thread emulation into a file of its own

p99/p99_threads.h
p99/p99_threads_posix.h

commit 8ff0c8795157357b97b0bde0ecc333a399063112
Author: Jens Gustedt <Jens.Gustedt@inria.fr>
Date:   Thu Jul 31 11:01:37 2014 +0200

    p99_rand uses extension features of p99_once_flag, so once_flag is not
    sufficient

p99/p99_rand.h

commit e1e5d9e877a8a81ab65617f7a9986c9c5a069ddb
Author: Jens Gustedt <Jens.Gustedt@inria.fr>
Date:   Thu Jul 31 10:58:43 2014 +0200

    avoid an error of impossible inlining

p99/p99_notifier.h

commit c5a0faa29e4405dfdf489eccaf8964ab839e69ab
Author: Jens Gustedt <Jens.Gustedt@inria.fr>
Date:   Thu Jul 31 10:57:50 2014 +0200

    avoid a warning the the system independent futex code

p99/p99_futex_c11.h

commit 70a1e52168a763fdb52114baf8ecae956a24790f
Author: Jens Gustedt <Jens.Gustedt@inria.fr>
Date:   Thu Jul 31 10:57:12 2014 +0200

    work around glibc' bogus strerror_r function

p99/p99_constraint.h

commit 9c299c4cef6ba36d8c0618c678dfa95a80877210
Author: Jens Gustedt <Jens.Gustedt@inria.fr>
Date:   Thu Jul 31 10:55:59 2014 +0200

    finally we don't need volatile for some operations

p99/p99_atomic.h

commit 0256484d8837799ec6fd3abecad5ca36e9b14128
Author: Jens Gustedt <Jens.Gustedt@inria.fr>
Date:   Tue Jul 29 19:06:15 2014 +0200

    add a hack around the broken strerror_r function in glibc

p99/p99_clib.h

commit f60b22703a50c97c975f8c7ad920db1b9a22e638
Author: Jens Gustedt <Jens.Gustedt@inria.fr>
Date:   Thu Jul 24 23:53:15 2014 +0200

    only define missing C11 interfaces if we are on a pre-C11 platform

p99/p99_clib.h


p99/p99_id.h
/* This may look like nonsense, but it really is -*- mode: C -*-              */
/*                                                                            */
/* Except for parts copied from previous work and as explicitly stated below, */
/* the authors and copyright holders for this work are as follows:            */
/* (C) copyright  2010-2013 Jens Gustedt, INRIA, France                       */
/* (C) copyright  2012 William Morris                                         */
/*                                                                            */
/* This file is free software; it is part of the P99 project.                 */
/* You can redistribute it and/or modify it under the terms of the QPL as     */
/* given in the file LICENSE. It is distributed without any warranty;         */
/* without even the implied warranty of merchantability or fitness for a      */
/* particular purpose.                                                        */
/*                                                                            */
#ifndef     P99_ID_H_
# define    P99_ID_H_

#include "p99_if.h"

/**
 ** @file
 **
 ** @brief Implement a simple FILEID that changes each time this file
 ** is included somewhere
 **
 ** The FILEID here is a hexadecimal number with 4 digits. 34320
 ** different such numbers are produced by the algorithm until it
 ** wraps around.
 **/

#define P00_DOCUMENT_ID(X) /*! @brief counting the number of inclusions of "p99_id.h" mod X */

P00_DOCUMENT_ID(11)
#define P99_ID_11 0

P00_DOCUMENT_ID(13)
#define P99_ID_13 0

P00_DOCUMENT_ID(15)
#define P99_ID_15 0

P00_DOCUMENT_ID(16)
#define P99_ID_16 0

#define P00_ALPH_0 0
#define P00_ALPH_1 1
#define P00_ALPH_2 2
#define P00_ALPH_3 3
#define P00_ALPH_4 4
#define P00_ALPH_5 5
#define P00_ALPH_6 6
#define P00_ALPH_7 7
#define P00_ALPH_8 8
#define P00_ALPH_9 9
#define P00_ALPH_10 A
#define P00_ALPH_11 B
#define P00_ALPH_12 C
#define P00_ALPH_13 D
#define P00_ALPH_14 E
#define P00_ALPH_15 F
#define P00_ALPH_16 G
#define P00_ALPH_17 H
#define P00_ALPH_18 I
#define P00_ALPH_19 J
#define P00_ALPH_20 K
#define P00_ALPH_21 L
#define P00_ALPH_22 M
#define P00_ALPH_23 N
#define P00_ALPH_24 O
#define P00_ALPH_25 P
#define P00_ALPH_26 Q
#define P00_ALPH_27 R
#define P00_ALPH_28 S
#define P00_ALPH_29 T
#define P00_ALPH_30 U
#define P00_ALPH_31 V
#define P00_ALPH_32 W
#define P00_ALPH_33 X
#define P00_ALPH_34 Y
#define P00_ALPH_35 Z



#define P00_ALPH_(A, X) P99_PASTE2(A, X)
#define P00_ALPH(X) P00_ALPH_(P00_ALPH_, X)

#define P99_ID_11_ P00_ALPH(P99_ID_11)
#define P99_ID_13_ P00_ALPH(P99_ID_13)
#define P99_ID_15_ P00_ALPH(P99_ID_15)
#define P99_ID_16_ P00_ALPH(P99_ID_16)

#define P99_ID() P99_PASTE4(P99_ID_11_, P99_ID_13_, P99_ID_15_, P99_ID_16_)

/**
 ** @brief A number that identifies different times of inclusion of
 ** the file "p99_id.h"
 **
 ** This is an unsigned hexadecimal constant of at most 4 significant digits.
 **/
#define P99_FILENO() P99_PASTE2(0x, P99_ID(), U)

/**
 ** @brief An identifier token that identifies different times of
 ** inclusion of the file "p99_id.h"
 **
 ** The argument may contain a list of partial identifier tokens that
 ** are appended to the id.
 **/
#define P99_FILEID(...) P99_PASTID(p00_fileid, P99_ID(), __VA_ARGS__)

/**
 ** @brief A number that identifies different lines in an include
 ** hierarchy.
 **
 ** This is a hexadecimal constant of type @c unsigned long long.
 ** @see P99_LINEID
 **/
#define P99_LINENO() P99_PASTE4(0x, __LINE__, P99_ID(), ULL)
/**
 ** @brief An identifier token that identifies different lines in an include
 ** hierarchy.
 **
 ** The argument may contain a token that is appended to the id.
 **
 ** If during an individual compilation there is a conflict because
 ** this macro is used in lines with exactly the same __LINE__ number,
 ** you'd have to include the file "p99_id.h" once more in one of the
 ** corresponding header files.
 **
 ** If such a conflict only occurs when linking different object files
 ** to your executable you should reconsider the offending lines that
 ** use this macro more thoroughly.
 **/
#define P99_LINEID(...) P99_PASTID(p00_lineid, __LINE__, P99_ID(), __VA_ARGS__)

#ifndef P00_UNIQ
# if __COUNTER__ || __COUNTER__
#  define P00_UNIQ __COUNTER__
# else
#  define P00_UNIQ p00_no_counter_feature
# endif
#endif

/**
 ** @brief An identifier token that is intended to be unique in this
 ** compilation unit
 **
 ** This uses different line numbers in an include hierarchy to
 ** disambiguate, and, if available, the non-standard macro
 ** <code>__COUNTER__</code>.
 **
 ** The argument may contain a token that is appended to the id.
 **
 ** If during an individual compilation there is a conflict because
 ** this macro is used in lines with exactly the same __LINE__ number,
 ** and <code>__COUNTER__</code> is not available you'd have to
 ** include the file "p99_id.h" once more in one of the corresponding
 ** header files.
 **
 ** If you need the same unique identifier at several places in the
 ** same macro you'd have to cascade as in the following example:
 **
 ** @code
 ** #define MY_DO(ID, START, STOP, INC) MY_DO_(ID, P99_UNIQ(n), P99_UNIQ(p), (START), (STOP), (INC))
 ** #define MY_DO_(ID, NU, PU, START, STOP, INC) for (unsigned ID = START, NU = STOP, PU = INC; ID < NU; ID += PU)
 **
 ** MY_DO(i, 0, strlen(filename), 1)
 **   printf("%c-", filename[i]);
 ** @endcode
 **
 ** This would expand to
 **
 ** @code
 ** for (unsigned ID = 0, crude_n = strlen(filename), crude_p = 1; i < crude_n; i += crude_p)
 **    printf("%c-", filename[i]);
 ** @endcode
 **
 ** Here, in the top most expansion of <code>MY_DO</code> two
 ** identifiers are generated and passed along as arguments of
 ** <code>MY_DO_</code>. (The "names" @c crude_n or @c crude_p only
 ** stand for some crude unique identifier you don't really want to
 ** know.)
 **
 ** The effect here is that the generated <code>for</code> loop has
 ** three local variables, that are initialized to the three arguments
 ** of <code>MY_DO</code>. By this, this macro can guarantee that the
 ** loop bounds are only evaluated once and that the number of
 ** iterations that is performed is known on entry to the loop.
 **/
#define P99_UNIQ(...)  P99_IF_EMPTY(__VA_ARGS__)(P99_LINEID(P00_UNIQ, uniq))(P99_LINEID(P00_UNIQ, uniq, __VA_ARGS__))

#ifdef P00_DOXYGEN
/**
 ** @brief Advance the unique ID generated by P99
 **
 ** Use this as
 ** @code
 ** #include P99_ADVANCE_ID
 ** @endcode
 **/
# define P99_ADVANCE_ID P99_ADVANCE_ID
#else
# define P99_ADVANCE_ID "p99_id.h"
#endif

#endif


#if P99_ID_11 >= 10
# undef P99_ID_11
# define P99_ID_11 0
#elif P99_ID_11 == 0
# undef P99_ID_11
# define P99_ID_11 1
#elif P99_ID_11 == 1
# undef P99_ID_11
# define P99_ID_11 2
#elif P99_ID_11 == 2
# undef P99_ID_11
# define P99_ID_11 3
#elif P99_ID_11 == 3
# undef P99_ID_11
# define P99_ID_11 4
#elif P99_ID_11 == 4
# undef P99_ID_11
# define P99_ID_11 5
#elif P99_ID_11 == 5
# undef P99_ID_11
# define P99_ID_11 6
#elif P99_ID_11 == 6
# undef P99_ID_11
# define P99_ID_11 7
#elif P99_ID_11 == 7
# undef P99_ID_11
# define P99_ID_11 8
#elif P99_ID_11 == 8
# undef P99_ID_11
# define P99_ID_11 9
#elif P99_ID_11 == 9
# undef P99_ID_11
# define P99_ID_11 10
#elif P99_ID_11 == 10
# undef P99_ID_11
# define P99_ID_11 11
#elif P99_ID_11 == 11
# undef P99_ID_11
# define P99_ID_11 12
#elif P99_ID_11 == 12
# undef P99_ID_11
# define P99_ID_11 13
#elif P99_ID_11 == 13
# undef P99_ID_11
# define P99_ID_11 14
#elif P99_ID_11 == 14
# undef P99_ID_11
# define P99_ID_11 15
#elif P99_ID_11 == 15
# undef P99_ID_11
# define P99_ID_11 16
#elif P99_ID_11 == 16
# undef P99_ID_11
# define P99_ID_11 17
#elif P99_ID_11 == 17
# undef P99_ID_11
# define P99_ID_11 18
#elif P99_ID_11 == 18
# undef P99_ID_11
# define P99_ID_11 19
#elif P99_ID_11 == 19
# undef P99_ID_11
# define P99_ID_11 20
#endif

#if P99_ID_13 >= 12
# undef P99_ID_13
# define P99_ID_13 0
#elif P99_ID_13 == 0
# undef P99_ID_13
# define P99_ID_13 1
#elif P99_ID_13 == 1
# undef P99_ID_13
# define P99_ID_13 2
#elif P99_ID_13 == 2
# undef P99_ID_13
# define P99_ID_13 3
#elif P99_ID_13 == 3
# undef P99_ID_13
# define P99_ID_13 4
#elif P99_ID_13 == 4
# undef P99_ID_13
# define P99_ID_13 5
#elif P99_ID_13 == 5
# undef P99_ID_13
# define P99_ID_13 6
#elif P99_ID_13 == 6
# undef P99_ID_13
# define P99_ID_13 7
#elif P99_ID_13 == 7
# undef P99_ID_13
# define P99_ID_13 8
#elif P99_ID_13 == 8
# undef P99_ID_13
# define P99_ID_13 9
#elif P99_ID_13 == 9
# undef P99_ID_13
# define P99_ID_13 10
#elif P99_ID_13 == 10
# undef P99_ID_13
# define P99_ID_13 11
#elif P99_ID_13 == 11
# undef P99_ID_13
# define P99_ID_13 12
#elif P99_ID_13 == 12
# undef P99_ID_13
# define P99_ID_13 13
#elif P99_ID_13 == 13
# undef P99_ID_13
# define P99_ID_13 14
#elif P99_ID_13 == 14
# undef P99_ID_13
# define P99_ID_13 15
#elif P99_ID_13 == 15
# undef P99_ID_13
# define P99_ID_13 16
#elif P99_ID_13 == 16
# undef P99_ID_13
# define P99_ID_13 17
#elif P99_ID_13 == 17
# undef P99_ID_13
# define P99_ID_13 18
#elif P99_ID_13 == 18
# undef P99_ID_13
# define P99_ID_13 19
#elif P99_ID_13 == 19
# undef P99_ID_13
# define P99_ID_13 20
#endif

#if P99_ID_15 >= 14
# undef P99_ID_15
# define P99_ID_15 0
#elif P99_ID_15 == 0
# undef P99_ID_15
# define P99_ID_15 1
#elif P99_ID_15 == 1
# undef P99_ID_15
# define P99_ID_15 2
#elif P99_ID_15 == 2
# undef P99_ID_15
# define P99_ID_15 3
#elif P99_ID_15 == 3
# undef P99_ID_15
# define P99_ID_15 4
#elif P99_ID_15 == 4
# undef P99_ID_15
# define P99_ID_15 5
#elif P99_ID_15 == 5
# undef P99_ID_15
# define P99_ID_15 6
#elif P99_ID_15 == 6
# undef P99_ID_15
# define P99_ID_15 7
#elif P99_ID_15 == 7
# undef P99_ID_15
# define P99_ID_15 8
#elif P99_ID_15 == 8
# undef P99_ID_15
# define P99_ID_15 9
#elif P99_ID_15 == 9
# undef P99_ID_15
# define P99_ID_15 10
#elif P99_ID_15 == 10
# undef P99_ID_15
# define P99_ID_15 11
#elif P99_ID_15 == 11
# undef P99_ID_15
# define P99_ID_15 12
#elif P99_ID_15 == 12
# undef P99_ID_15
# define P99_ID_15 13
#elif P99_ID_15 == 13
# undef P99_ID_15
# define P99_ID_15 14
#elif P99_ID_15 == 14
# undef P99_ID_15
# define P99_ID_15 15
#elif P99_ID_15 == 15
# undef P99_ID_15
# define P99_ID_15 16
#elif P99_ID_15 == 16
# undef P99_ID_15
# define P99_ID_15 17
#elif P99_ID_15 == 17
# undef P99_ID_15
# define P99_ID_15 18
#elif P99_ID_15 == 18
# undef P99_ID_15
# define P99_ID_15 19
#elif P99_ID_15 == 19
# undef P99_ID_15
# define P99_ID_15 20
#endif

#if P99_ID_16 >= 15
# undef P99_ID_16
# define P99_ID_16 0
#elif P99_ID_16 == 0
# undef P99_ID_16
# define P99_ID_16 1
#elif P99_ID_16 == 1
# undef P99_ID_16
# define P99_ID_16 2
#elif P99_ID_16 == 2
# undef P99_ID_16
# define P99_ID_16 3
#elif P99_ID_16 == 3
# undef P99_ID_16
# define P99_ID_16 4
#elif P99_ID_16 == 4
# undef P99_ID_16
# define P99_ID_16 5
#elif P99_ID_16 == 5
# undef P99_ID_16
# define P99_ID_16 6
#elif P99_ID_16 == 6
# undef P99_ID_16
# define P99_ID_16 7
#elif P99_ID_16 == 7
# undef P99_ID_16
# define P99_ID_16 8
#elif P99_ID_16 == 8
# undef P99_ID_16
# define P99_ID_16 9
#elif P99_ID_16 == 9
# undef P99_ID_16
# define P99_ID_16 10
#elif P99_ID_16 == 10
# undef P99_ID_16
# define P99_ID_16 11
#elif P99_ID_16 == 11
# undef P99_ID_16
# define P99_ID_16 12
#elif P99_ID_16 == 12
# undef P99_ID_16
# define P99_ID_16 13
#elif P99_ID_16 == 13
# undef P99_ID_16
# define P99_ID_16 14
#elif P99_ID_16 == 14
# undef P99_ID_16
# define P99_ID_16 15
#elif P99_ID_16 == 15
# undef P99_ID_16
# define P99_ID_16 16
#elif P99_ID_16 == 16
# undef P99_ID_16
# define P99_ID_16 17
#elif P99_ID_16 == 17
# undef P99_ID_16
# define P99_ID_16 18
#elif P99_ID_16 == 18
# undef P99_ID_16
# define P99_ID_16 19
#elif P99_ID_16 == 19
# undef P99_ID_16
# define P99_ID_16 20
#endif
